core/App/ehsm_napi.cpp

```diff
@@ -1125,4 +1129,115 @@ char* NAPI_ExportDataKey(const char* cmk_base64,
     return retJsonObj.toChar();
 }
 
-}  // extern "C"
+/*
+@return
+[string] json string
+    {
+        code: int,
+        message: string,
+        result: {
+            challenge_base64 : string,
+            ga_base64 : string
+        }
+    }
+*/
+char *NAPI_RA_HANDSHAKE_MSG0(const char *p_msg0)
+{
+    log_d("***NAPI_RA_HANDSHAKE_MSG0 start.");
+
+    log_d("msg0: \n %s", p_msg0);
+    std::string challenge_base64 = "challenge_base64";
+    std::string ga_base64 = "ga_base64";
+
+    RetJsonObj retJsonObj;
+    
+    retJsonObj.addData_string("challenge_base64", challenge_base64);
+    retJsonObj.addData_string("ga_base64", ga_base64);
+    log_d("msg1: \n%s",retJsonObj.toChar());
+
+    log_d("***NAPI_RA_HANDSHAKE_MSG0 end.");
+    return retJsonObj.toChar();
+}
+
+/*
+@return
+[string] json string
+    {
+        code: int,
+        message: string,
+        result: {
+            msg3_base64 : string
+        }
+    }
+*/
+char *NAPI_RA_HANDSHAKE_MSG2(const char *p_msg2)
+{
+    log_d("***NAPI_RA_HANDSHAKE_MSG2 start.");
+
+    log_d("msg2: \n %s", p_msg2);
+    std::string msg3_base64 = "msg3_base64";
+
+    RetJsonObj retJsonObj;
+    retJsonObj.addData_string("msg3_base64", msg3_base64);
+    log_d("msg3: \n%s",retJsonObj.toChar());
+
+    log_d("***NAPI_RA_HANDSHAKE_MSG2 end.");
+    return retJsonObj.toChar();
+}
+
+/*
+@return
+[string] json string
+    {
+        code: int,
+        message: string,
+        result: {
+            appid : string
+            apikey : string
+        }
+    }
+*/
+char *NAPI_RA_GET_API_KEY(const char *p_msg4)
+{
+    log_d("***NAPI_RA_GET_API_KEY start.");
+    ehsm_status_t ret = EH_OK;
+    RetJsonObj retJsonObj;
+
+    ehsm_data_t apikey_data;
+    char *p_appid;
+
+    memset(&apikey_data, 0, sizeof(apikey_data));
+
+    log_d("msg4: \n %s", p_msg4);
+
+    // create appid
+    uuid_t uu;
+    uuid_generate(uu);
+    p_appid = (char*)malloc(37);
```

Comment:
37 -> sizeof(uuid_t) + 1

Reply:
prefer to define as:
char p_appid[sizeof(uuid_t) + 1] = {0};
malloc does not zero the last byte. You have to set it by self or use calloc().